dev/core#2122 - Upgrade scripts for revert event timezones

demeritcowboy opened the pull request:
Overview
----------------------------------------
This goes along with PR https://github.com/civicrm/civicrm-core/pull/22930 and is not really standalone, but I've separated it out since it is the most likely part to get merge-conflicted or need some tweaking.

Note the steps are re-runnable since it won't do anything if the field is datetime. For this to all work it also needs a backport of the 5.47 parts of this PR into the 5.47.3 branch. Then:

* <5.47 to 5.47.3 - it would NOT run the 5.47.3 backport script because the field will still be datetime because the other PR removes the 5.47.alpha1 steps. It's like nothing ever happened.
* <5.47 to 5.48 - it would run neither script for the same reasons
* 5.47.1 to 5.47.3 - it would run the 5.47.3 script
* 5.47.1 to 5.48 - it would run the 5.47.3 script and NOT run the 5.48.beta1 script because when it hits the 5.48 script the field will be datetime.
* 5.47.3 to 5.48 - it would not run the 5.48.beta1 script because the field will be datetime at that point (for both upgrades and new installs of 5.47.3)

I've done some testing. To realistically test the "upgrade-x-to-5.47.3-and-stop-there" path you would use the 5.47 backport at https://github.com/civicrm/civicrm-core/pull/22946

Before
----------------------------------------


After
----------------------------------------


Technical Details
----------------------------------------
See https://lab.civicrm.org/dev/core/-/issues/2122

Comments
----------------------------------------
If you had upgraded to 5.47, then for troubleshooting purposes it creates some backup fields with what the values were before "fixing", as suggested in chat.

demeritcowboy commented:
I'm not worried about the test fail because this isn't meant to be standalone. This PR goes with the other PR, so the upgrade test shouldn't even hit the line where it's failing.

demeritcowboy commented:
One thing that has come up: If the drupal user timezone (or cms timezone for wordpress) of the user account that ran the original upgrade to 5.47 is different from the timezone of the user account running the revert, the times could be even more off when reverted. Thinking...

KarinG commented:
Dave and I tested the revert on one of my live site and unintentionally switched the user who ran the upgrade ->

- before 5.47 -> event start time was 1pm (MST)
- after upgrade -> I ran it with my Drupal CMS in UTC (I was testing our fullcalendar integration - had set it to UTC)  -> so it became -> event start time was 1pm UTC
- after revert -> the start time now is 7am -> after some thinking -> yes the user running the revert today was in Edmonton TZ - that's UTC-6 since we switched yesterday 7 + 6 = 13:00 -> 1pm

So all ✅

So we inadvertently tested the "what if the user performing the upgrade moves from UTC to Edmonton after the original upgrade to 5.47.* but before the revert" scenario ✅

Why are checks failing?

demeritcowboy commented:
I mentioned the fail at https://github.com/civicrm/civicrm-core/pull/22940#issuecomment-1066224516. This isn't a standalone PR, just separated out to make conflict resolution/updating easier.

KarinG commented:
Checking the backup now ->

KarinG commented:
That looks good -> 

<img width="1442" alt="image" src="https://user-images.githubusercontent.com/5340555/158290668-6729a071-0bde-464c-9f64-66219dbbf632.png">

agileware-fj commented:
Have tested and can confirm that the timestamp column:

1. is correct for events in the same timezone offset as the CRM at the time the upgrade is run.
2. is incorrect for events that have been saved in a different timezone (updated after the upgrade).  
3. works ok for events that were saved on the other side of a DST change, as long as that change has not eventuated in the mean time.

For point (2), I'm not sure that there's any way to sensibly deal with this on a MySQL instance where the timezone data has not been filled (and I suspect this applies to most systems).  In the cases where it *has*, it would be possible - assuming the MySQL session time zone is correct - to do something like
```sql
UPDATE civicrm_event SET start_date = CONVERT_TZ(start_date, @@session.time_zone, event_tz_bak)
                           end_date = CONVERT_TZ(end_date, @@session.time_zone, event_tz_bak)
            registration_start_date = CONVERT_TZ(registration_start_date, @@session.time_zone, event_tz_bak)
              registration_end_date = CONVERT_TZ(registration_end_date, @@session.time_zone, event_tz_bak);
```

Also for anyone else testing this, you might find the data clearer if you terminate your SQL statements with \G, e.g.
`SELECT start_date, start_date_ts_bak, end_date, end_date_ts_bak, registration_start_date, registration_start_date_ts_bak, registration_end_date, registration_end_date_ts_bak, event_tz_bak FROM civicrm_event\G`, which gets you output like:
```
*************************** 9. row ***************************
                    start_date: 2022-03-18 13:00:00
             start_date_ts_bak: 2022-03-18 13:00:00
                      end_date: 2022-03-18 15:00:00
               end_date_ts_bak: 2022-03-18 15:00:00
       registration_start_date: NULL
registration_start_date_ts_bak: NULL
         registration_end_date: NULL
  registration_end_date_ts_bak: NULL
                  event_tz_bak: Australia/Sydney
```

demeritcowboy commented:
I've added pre and post messages that only display if you're upgrading from a path where the events were changed. The pre message works but post messages no longer seem to be working at all in core?

Pre message:
<img width="565" alt="Untitled2" src="https://user-images.githubusercontent.com/2967821/158420627-b4cc9e92-f50f-48ca-b0e1-eb1960c2d99f.png">

demeritcowboy commented:
So the post upgrade messages seems to be yet another issue in recent civi. It does work with `cv`, but not the UI.

I've been thinking about `start_date = CONVERT_TZ(start_date, @@session.time_zone, event_tz_bak)`. Besides the noted fact that named timezones are not installed on many mysql/mariadb installs:
* We're assuming this is run from inside civi, so session.time_zone is an offset. So then it would have to limit it to only ones you already know are shifted, otherwise it will also shift ones that are fine. It might be possible to get creation dates from civicrm_log or log_civicrm_event.
  * Noting also that log_civicrm_event.start_date is unfortunately not usable because it will also have been converted to timestamp once the schema reconciliation ran.

Thinking out loud, since the event ids haven't changed, if people have backups from before the upgrade to 5.47, and not too many events have been rescheduled in real life, then after the revert upgrade, doing the following might be a reasonable option, it's just not one we can run for them.
  * Restoring the backup civicrm_event table to a new temporary table tmp_civicrm_event
  * Running `UPDATE civicrm_event e INNER JOIN tmp_civicrm_event t ON t.id = e.id SET e.start_date = t.start_date;`
  * `UPDATE civicrm_event e INNER JOIN tmp_civicrm_event t ON t.id = e.id SET e.end_date = t.end_date;`
  * `UPDATE civicrm_event e INNER JOIN tmp_civicrm_event t ON t.id = e.id SET e.registration_start_date = t.registration_start_date;`
  * `UPDATE civicrm_event e INNER JOIN tmp_civicrm_event t ON t.id = e.id SET e.registration_end_date = t.registration_end_date;`

agh1 commented:
[Cross-referencing from Mattermost](https://chat.civicrm.org/civicrm/pl/5rdpjws3afbr8ed864txmrskae), could we please avoid merging this and #22930 until after the security release is cut? I think it will simplify messaging to keep things separate and avoid introducing mayhem into a security release.

If people are on 5.47.x, they're no worse off if the security release lacks the revert.  If they're on an earlier version, we can/should point out that 5.46.3 is preferred because of this timezone problem.

totten commented:
I've been trying this out with the combined/5.47 variant (#22946). A few things:

1. The upgrade message strikes me as quite informative, but obviously the subject-matter is apt for confusion. I think it would better to go with a thinner message - and put the explanatory content in a wiki-link (somewhere under https://lab.civicrm.org/dev/core/-/wikis/home). That way we can revise in response to issues/questions that come up post-release. (This could be a new wiki page like "CiviEvent v5.47 Downgrade Notice"? Or use the PR-description here to track the initial draft - and we can copy it into [the an existing page](https://lab.civicrm.org/dev/core/-/wikis/CiviEvent-v5.47-Timezone-Notice) when we release.)
2. I was concerned about what would happen to message-templates -- if they would wind up with an invalid token. Did a trial run of `5.46.3` => `5.47.2` => `5.47.3[PR]`:
    * It does leave an extra reference to `{$event.event_tz}`.
    * I performed an offline-registration on php72 and php80. In my configuration, both produced a warning ("Warning: Undefined array key "event_tz" in eval() (line 56 of /home/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php(1979) : eval()'d code)."). However, it was not fatal.
    * It's not entirely clear what one should do with these message-templates - eg (a) define a step to upgrade the msgtpls, or (b) define a step to strip out `{$event.event_tz}`, or (c) define a placeholder value `$smarty->assign(...event_tz => ''...)`.
        * If the behavior is going to get flip-flopped a few times...  then maybe the placeholder-value is better (so that the DB doesn't flip-flop; ie the live content in the DB can be used on different revisions without needing further DB updates)?
        * This maybe depends on one's expectation for "How much do these templates get customized?" I doubt many people have specifically targeted many customizations of the msgtpl's at this field. However, there's really no way to measure that...

demeritcowboy commented:
For the message templates, it seems I have the CRM/Upgrade/Incremental/MessageTemplates array in the 5.48 version but didn't add anything to the 5.47 PR. In theory I just need to add the array in that PR for 5.47.3.

demeritcowboy commented:
I've:
* Added to missing message template array to the 5.47 version and tested it and it updates the message templates.
* Shortened the pre/post upgrade messages.
* Added the link to the wiki in the messages.

demeritcowboy committed:
dev/core#2122 - Add TimezoneRevertTrait (shared by 5.47/5.48)

Based on upgrade-steps from demerit. Adapted to trait by totten.

totten committed:
Set version to 5.48.beta2

totten committed:
dev/core#2122 - For 5.48.beta2, use TimezoneRevertTrait and updates msg tpls

totten committed:
dev/core#2122 - TimezoneRevertTrait - Split out messaging across bullets

demeritcowboy commented on CRM/Upgrade/Incremental/php/TimezoneRevertTrait.php:
```diff
@@ -0,0 +1,118 @@
+   * of the account that was used to do the original upgrade to 5.47.
+   * @return string
+   */
+  private static function getMajorityTimezone(): string {
```
I had removed this function when I shortened the message. Unless the point of this trait is that you're expecting this is going to be needed again elsewhere. Hopefully we're not going to regularly need timezone revert features in future.

totten replied:
Yeah, apologies for the uncommented-push. 

I was `r-run`ning after your update yesterday and wanted to try tweaking the message again. A couple things coming out of that

(1) I found the messages were showing up twice because the code was duplicated in `FiveFortySeven` and `FiveFortyEight`. I figured this was an accident -- something easy to do when shuffling the patches back-and-forth between the two upgraders (`FiveFortySeven`/`FiveFortyEight`) and the two branches (`5.47`/`5.48`). The `TimezoneRevertTrait` can be an exact in both branches, so it should be easier to shuffle between them.

(2) I liked that the message was shorter, but it was funny - I would've trimmed the other way round. (eg I would've kept+trimmed different parts of your earlier text.) Giving some indication of the current CiviEvent seems useful. I think maybe it's better to use bullets or line-breaks (ie giving similar info - but in smaller chunks).

(3) Rather than get into literary analysis, I thought I'd play with the prose and send concrete suggestions. (But I went to bed before sending that up.) Will post that soon.

demeritcowboy commented on CRM/Upgrade/Incremental/php/FiveFortySeven.php:
```diff
@@ -69,10 +69,6 @@ public function upgrade_5_47_alpha1($rev): void {
       "timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP COMMENT 'Relationship last modified.'"
     );
     $this->addTask('Set initial value for relationship created_date and modified_date to start_date', 'updateRelationshipDates');
-    $this->addTask('core-issue#2122 - Add timezone column to Events', 'addColumn',
```
I didn't notice before that this duplicates the same removal in #22930, but I guess it works out the same in the end.

totten commented:
Related discussion on MM: https://chat.civicrm.org/civicrm/pl/gxy9tupkw3r9frxnd9ydfw3w6y

Rebased to include #22930. 

Continued tweaking the messaging:

* For a site which has all events in the same/default TZ:
    ![Screenshot from 2022-03-23 22-11-45](https://user-images.githubusercontent.com/1336047/159846739-87310f52-e380-4d8a-b285-19610b97e555.png)
* For a site which uses multiple TZs:
    ![Screenshot from 2022-03-23 22-14-36](https://user-images.githubusercontent.com/1336047/159847017-43bba317-b258-4fcc-8a41-3a99c8cdb739.png)

totten committed:
dev/core#2122 - TimezoneRevertTrait - Tweak messaging

totten committed:
dev/core#2122 - TimezoneRevertTrait - Downgrade should use same TZ as prior upgrade

totten committed:
dev/core#2122 - TimezoneRevertTrait - Cleanup modified TZ-offsets and DST skews

totten committed:
dev/core#2122 - TimezoneRevertTrait - Failsafe to opt-out of TZ converter

totten commented:
@demeritcowboy and I had some more discussion on MM about erroneous outputs (ie values that weren't converted well). Two major scenarios came up:

* _Passive scenario_: You have an upgrade (5.45 => 5.47.2), leave the data alone, and then downgrade (5.47.2 => 5.47.3 or 5.48.beta2). The values could come out distorted because the upgrade+downgrade steps were executing in different DSNs with different `@@time_zone`s. This one should be resolved by 9c0712453e79b66073936251343fa9fdc5595d5b.
* _Active web-user scenario_: You have an upgrade (5.45 => 5.47.2) and starting events (fixing times, filling-in timezones) -- then downgrade (5.47.2 => 5.47.3 or 5.48.beta2). The values could come out distorted because of the DST-bug and because of the varying `event_tz`s. The `event_tz`s are solvable, but the DST-bug is fairly tricky to unravel perfectly. Never-the-less, 24530df670b8863cc10b6190f00454fdce5b0c75 seems to be moderately effective heuristic.

To test these out, I used a moderately aggressive `r-run` procedure (https://gist.github.com/totten/c676e04fba28af61b994a77be88a8d1e) with a range of dates+TZs (chosen to exacerbate DST issues) and a mix of the active/passive usage above. Without these two commits, it did quite badly:

```
 +----+-----------------------------------------+---------------------+
 | id | title                                   | start_date          |
 +----+-----------------------------------------+---------------------+
 |  1 | Fall Fundraiser Dinner                  | 2022-07-05 20:00:00 | wrong
 |  2 | Summer Solstice Festival Day Concert    | 2022-01-04 16:00:00 | wrong
 |  3 | Rain-forest Cup Youth Soccer Tournament | 2022-08-05 10:00:00 | wrong
 |  4 | NULL                                    | NULL                | ok
 |  5 | NULL                                    | NULL                | ok
 |  6 | NULL                                    | NULL                | ok
 |  7 | First 1pm LA                            | 2022-03-01 16:00:00 | wrong
 |  8 | First 1pm NY                            | 2022-03-01 13:01:00 | ok
 |  9 | First 1pm HK                            | 2022-03-01 06:02:00 | wrong
 | 10 | Sixteen 2pm LA                          | 2022-03-16 17:00:00 | wrong
 | 11 | Sixteen 2pm NY                          | 2022-03-16 14:01:00 | ok
 | 12 | Sixteen 2pm HKI                         | 2022-03-16 08:02:00 | wrong
 | 13 | Thirtieth 3pm LA                        | 2022-03-30 19:00:00 | wrong
 | 14 | Thirtieth 3pm NY                        | 2022-03-30 16:01:00 | wrong
 | 15 | Thirtieth 3pm HKI                       | 2022-03-30 09:02:00 | wrong
 | 16 | April Fools Midnight                    | 2022-03-31 09:03:00 | wrong
 +----+-----------------------------------------+---------------------+
```

With the patches, it does better:

```
+----+-----------------------------------------+---------------------+
| id | title                                   | start_date          |
+----+-----------------------------------------+---------------------+
|  1 | Fall Fundraiser Dinner                  | 2022-07-05 17:00:00 |
|  2 | Summer Solstice Festival Day Concert    | 2022-01-04 12:00:00 |
|  3 | Rain-forest Cup Youth Soccer Tournament | 2022-08-05 07:00:00 |
|  4 | NULL                                    | NULL                |
|  5 | NULL                                    | NULL                |
|  6 | NULL                                    | NULL                |
|  7 | First 1pm LA                            | 2022-03-01 13:00:00 |
|  8 | First 1pm NY                            | 2022-03-01 13:01:00 |
|  9 | First 1pm HK                            | 2022-03-01 13:02:00 |
| 10 | Sixteen 2pm LA                          | 2022-03-16 14:00:00 |
| 11 | Sixteen 2pm NY                          | 2022-03-16 14:01:00 |
| 12 | Sixteen 2pm HKI                         | 2022-03-16 14:02:00 |
| 13 | Thirtieth 3pm LA                        | 2022-03-30 16:00:00 | off by 1, DST error
| 14 | Thirtieth 3pm NY                        | 2022-03-30 16:01:00 | off by 1, DST error
| 15 | Thirtieth 3pm HKI                       | 2022-03-30 16:02:00 | off by 1, DST error
| 16 | April Fools Midnight                    | 2022-04-01 00:03:00 |
+----+-----------------------------------------+---------------------+
```

The last few errors are not surprising - the CiviEvent admin user for `#13-16`  had a different DST policy from the upgrader. 

I had a theory about how to re-run this in a way that would swap DST policy (eg if your sysadmin-user happened to be in an EU but the regular Civi users were in US, then you'd re-run with a hint to use US DST scheduling). However, this theory part hasn't panned out yet.

demeritcowboy commented on CRM/Upgrade/Incremental/php/TimezoneRevertTrait.php:
```diff
@@ -0,0 +1,302 @@
+    $events = CRM_Utils_SQL_Select::from('civicrm_event')
+      ->select(['id', 'event_tz_bak'])
+      ->select(array_values($columns))
+      ->where('id in (#IDS)', ['IDS' => array_keys($eventModTimes)])
```
I get a syntax error here during one of my tests, I presume because I didn't edit any events after upgrading to 5.47.2.

`SELECT id, event_tz_bak, start_date_ts_bak, end_date_ts_bak, registration_start_date_ts_bak, registration_end_date_ts_bak FROM civicrm_event WHERE (id in ()) [nativecode=1064 ** You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '))' at line 3]`

totten replied:
Yup, I can reproduce that. Fixed.

totten commented:
Merging based on MM discussion.
